File: frontend/src/types/knowledge.ts
export type KBScope = 'mine' | 'team' | 'verified' | 'reference'

// What the "Optimized" chip means for a KB, from the v2 list / detail endpoint.
//   applied   — tuned RAG settings are live in chat
//   stale     — live, but sources or test questions changed materially since tuning
//   available — a completed optimization has settings that were never applied
export interface KBOptimizationStatus {
  state: 'applied' | 'stale' | 'available'
  applied_at?: string | null
  applied_run_uuid?: string | null
  last_run_at?: string | null
  last_run_uuid?: string | null
  tuned_keys?: string[]
  stale: boolean
  stale_reasons: string[]
  sources_at_run?: number
  sources_added?: number
  sources_removed?: number
  queries_at_run?: number
  queries_added?: number
  queries_removed?: number
  queries_edited?: number
}

export interface KnowledgeBase {
  uuid: string
  title: string
  description: string
  // 'unavailable' appears only on reference rows whose source KB no longer
  // resolves (deleted, retired from the catalog, or no longer shared).
  status: 'empty' | 'building' | 'ready' | 'error' | 'unavailable'
  shared_with_team: boolean
  team_owned: boolean
  verified: boolean
  organization_ids: string[]
  tags: string[]
  team_id: string | null
  total_sources: number
  sources_ready: number
  sources_failed: number
  total_chunks: number
  created_at: string
  updated_at: string
  // Scope & ownership fields (from v2 list endpoint)
  user_id?: string
  scope?: KBScope
  is_reference?: boolean
  source_kb_uuid?: string
  reference_uuid?: string
  // Whether the current user may run manage-level actions (add sources, rename,
  // share, delete). False for e.g. an adopted verified catalog KB the user
  // doesn't own. Treat a missing value as true — read it as `can_manage !== false`.
  can_manage?: boolean
  // Set by KB Autovalidate's apply path
  has_optimized_config?: boolean
  optimized_config_set_at?: string | null
  // Full story behind the chip (see KBOptimizationStatus); null when nothing to say.
  optimization?: KBOptimizationStatus | null
  // AI-trust signals from the latest validation run. Scores are 0-1.
  last_validation_score?: number | null
  last_validation_baseline_score?: number | null
  last_validation_lift?: number | null
  last_validated_at?: string | null
  // Per-requesting-user: when this user last chatted with the KB (ISO string).
  last_used_at?: string | null
}

/**
 * Refresh / ingestion provenance for one source (backend
 * app/utils/kb_source_currency.py). Lets an evaluator verify source currency
 * without re-fetching the original.
 */
export interface SourceCurrency {
  status: 'never_ingested' | 'ingested' | 'refreshed' | 'unchanged' | 'retained_previous' | 'retrieval_failed' | 'ingestion_failed'
  last_refresh_attempted_at: string | null
  last_retrieved_at: string | null
  last_ingested_at: string | null
  // When the text currently held and served was retrieved — a failed refresh
  // leaves it alone, so this is the "retained content" date.
  content_retrieved_at: string | null
  content_hash: string | null
  content_hash_algorithm: string
  // false: computed from the retained snapshot because the source predates
  // hash recording at ingest.
  content_hash_recorded: boolean
  last_refresh_outcome: 'refreshed' | 'unchanged' | 'retrieval_failed' | 'ingestion_failed' | null
  last_refresh_error: string | null
}

export interface KnowledgeBaseSource {
  uuid: string
  source_type: 'document' | 'url'
  document_uuid?: string
  document_title?: string
  // False when the source's document has been deleted from Files. The KB still
  // answers from the chunks it indexed, so the source is labelled, not hidden.
  document_exists?: boolean | null
  url?: string
  url_title?: string
  custom_name?: string | null
  source_reference?: string | null
  status: 'pending' | 'processing' | 'ready' | 'error'
  error_message?: string
  chunk_count: number
  // URL source whose extracted text was cut off at the fetcher size cap:
  // "ready" but incomplete, so the UI warns instead of showing a clean check.
  truncated?: boolean
  created_at: string
  // When the text was last fetched/ingested (null while pending).
  processed_at?: string | null
  currency?: SourceCurrency | null
}

export interface KnowledgeBaseSourceDetail extends KnowledgeBaseSource {
  content?: string | null
  crawl_enabled: boolean
  max_crawl_pages: number
  parent_source_uuid?: string | null
  crawled_urls?: string[] | null
  // Navigation pages the crawl followed for their links but did not embed.
  skipped_urls?: string[] | null
  child_sources: KnowledgeBaseSource[]
}

export interface KnowledgeBaseDetail extends KnowledgeBase {
  sources: KnowledgeBaseSource[]
}

export interface KBListResponse {
  items: KnowledgeBase[]
  total: number
}

export interface KBReference {
  uuid: string
  source_kb_uuid: string
  user_id: string
  team_id?: string
  note?: string
  pinned: boolean
  created_at?: string
}

File: backend/app/schemas/knowledge.py
"""Knowledge Base schemas for request/response validation."""

from typing import Optional

from pydantic import BaseModel

from app.utils.naming import EntityName, OptionalEntityName


class CreateKBRequest(BaseModel):
    title: EntityName
    description: Optional[str] = None


class UpdateKBRequest(BaseModel):
    title: OptionalEntityName = None
    description: Optional[str] = None
    shared_with_team: Optional[bool] = None
    organization_ids: Optional[list[str]] = None
    tags: Optional[list[str]] = None


class AddDocumentsRequest(BaseModel):
    document_uuids: list[str] = []
    folder_uuids: list[str] = []


class AddFolderRequest(BaseModel):
    folder_uuid: str
    include_subfolders: bool = True


class ConvertDocumentsRequest(BaseModel):
    """Wrap one or more SmartDocuments in a new KB so they can be retrieved
    instead of inlined. Used by the chat / workflow "Convert to Knowledge
    Base" affordance shown when a doc is too large for the current model.
    """
    document_uuids: list[str]
    title: Optional[str] = None  # defaults to the first doc's title


class ShareKBRequest(BaseModel):
    comment: Optional[str] = None


class AddUrlsRequest(BaseModel):
    urls: list[str]
    crawl_enabled: bool = False
    max_crawl_pages: int = 5
    allowed_domains: str = ""  # comma-separated hosts, optionally with path prefixes (example.com/irb)


class KBSourceCurrency(BaseModel):
    """Refresh / ingestion provenance for one source (see
    app.utils.kb_source_currency). Lets an evaluator verify source currency
    from the UI or the export without re-fetching every original."""

    # never_ingested | ingested | refreshed | unchanged | retained_previous
    # | retrieval_failed | ingestion_failed
    status: str
    last_refresh_attempted_at: Optional[str] = None
    last_retrieved_at: Optional[str] = None
    last_ingested_at: Optional[str] = None
    # When the text currently held and served was retrieved — unchanged by a
    # failed refresh, so it is the "retained content" date.
    content_retrieved_at: Optional[str] = None
    content_hash: Optional[str] = None
    content_hash_algorithm: str = "sha256"
    # False when the hash was computed just now from the retained snapshot
    # because the source predates hash recording at ingest.
    content_hash_recorded: bool = False
    last_refresh_outcome: Optional[str] = None
    last_refresh_error: Optional[str] = None


class KBSourceResponse(BaseModel):
    uuid: str
    source_type: str
    document_uuid: Optional[str] = None
    document_title: Optional[str] = None  # Resolved from SmartDocument for display
    url: Optional[str] = None
    url_title: Optional[str] = None
    custom_name: Optional[str] = None  # user-provided label; UI prefers this over title/url
    source_reference: Optional[str] = None  # user-verifiable provenance, shown as "Source: …"
    status: str
    error_message: Optional[str] = None
    chunk_count: int = 0
    # URL source whose extracted text was cut off at the fetcher size cap:
    # "ready" but incomplete, so the UI warns instead of showing a clean check.
    truncated: bool = False
    created_at: Optional[str] = None
    # When the source's text was last fetched/ingested. Surfaced on the list
    # so a user can tell how stale a URL snapshot is before refreshing it.
    processed_at: Optional[str] = None
    currency: Optional[KBSourceCurrency] = None


class KBSourceDetailResponse(KBSourceResponse):
    """Full source detail for the source inspector modal.

    Includes cached content (for URLs), crawl metadata, and references to
    parent/child sources when applicable.
    """

    content: Optional[str] = None  # Cached extracted text (URL sources)
    crawl_enabled: bool = False
    max_crawl_pages: int = 5
    parent_source_uuid: Optional[str] = None
    crawled_urls: Optional[list[str]] = None
    # Navigation pages the crawl followed for links but did not embed
    skipped_urls: Optional[list[str]] = None
    child_sources: list[KBSourceResponse] = []  # Crawled children (when this is a parent)


class UpdateSourceRequest(BaseModel):
    """Patch a single KB source. Only fields explicitly present are applied;
    an empty string clears that field (reverts to the auto-derived value)."""
    custom_name: Optional[str] = None
    source_reference: Optional[str] = None


class KBOptimizationStatusResponse(BaseModel):
    """What the "Optimized" chip means for this KB, and whether it still holds.

    ``applied``: tuned RAG settings are live. ``stale``: live, but the sources
    or test questions have changed materially since they were tuned.
    ``available``: a completed optimization has settings that were never
    applied (or were reverted). See ``services/kb_optimization_status``.
    """
    state: str  # applied | stale | available
    applied_at: Optional[str] = None
    applied_run_uuid: Optional[str] = None
    last_run_at: Optional[str] = None
    last_run_uuid: Optional[str] = None
    tuned_keys: list[str] = []
    stale: bool = False
    stale_reasons: list[str] = []
    sources_at_run: int = 0
    sources_added: int = 0
    sources_removed: int = 0
    queries_at_run: int = 0
    queries_added: int = 0
    queries_removed: int = 0
    queries_edited: int = 0


class KBResponse(BaseModel):
    uuid: str
    title: str
    description: Optional[str] = None
    status: str
    shared_with_team: bool = False
    team_owned: bool = False
    verified: bool = False
    organization_ids: list[str] = []
    tags: list[str] = []
    total_sources: int = 0
    sources_ready: int = 0
    sources_failed: int = 0
    total_chunks: int = 0
    created_at: Optional[str] = None
    updated_at: Optional[str] = None
    # Scope & ownership fields for the UI
    user_id: Optional[str] = None
    scope: Optional[str] = None  # "mine" | "team" | "verified" | "reference"
    is_reference: bool = False
    source_kb_uuid: Optional[str] = None  # set when is_reference=True
    reference_uuid: Optional[str] = None  # the reference's own uuid
    # Whether the requesting user may perform manage-level actions (add sources,
    # rename, share, delete). Lets the UI disable those affordances up front
    # instead of letting the user complete a flow that ends in a 403 — e.g. an
    # adopted verified catalog KB, which is viewable by everyone but manageable
    # only by its owner, an examiner, or an admin. Defaults True because the
    # endpoints that don't set it explicitly (create/import/adopt) only ever
    # return a KB the requester just became the owner of.
    can_manage: bool = True
    # Set by KB Autovalidate's apply path. Presence (not value) is what the UI
    # surfaces as a small "Optimized" chip.
    has_optimized_config: bool = False
    optimized_config_set_at: Optional[str] = None
    # Full story behind the chip — applied / stale / available, when, from
    # which run, and what changed since. None when there is nothing to say.
    # Populated by the v2 list and the detail endpoint.
    optimization: Optional[KBOptimizationStatusResponse] = None
    # AI-trust signals from the latest KB validation run.
    # Scores are 0-1; lift is also 0-1 (e.g., 0.28 == +28pts vs. baseline).
    last_validation_score: Optional[float] = None
    last_validation_baseline_score: Optional[float] = None
    last_validation_lift: Optional[float] = None
    last_validated_at: Optional[str] = None
    # Per-requesting-user: when this user last chatted with the KB. Powers the
    # "Recently Used" sort. None = never used (or legacy pre-tracking usage).
    last_used_at: Optional[str] = None


class KBListResponse(BaseModel):
    items: list[KBResponse] = []
    total: int = 0


class AdoptKBRequest(BaseModel):
    note: Optional[str] = None
    team_id: Optional[str] = None  # adopt to a specific team (default: personal)


class KBReferenceResponse(BaseModel):
    uuid: str
    source_kb_uuid: str
    user_id: str
    team_id: Optional[str] = None
    note: Optional[str] = None
    pinned: bool = False
    created_at: Optional[str] = None


class KBDetailResponse(KBResponse):
    sources: list[KBSourceResponse] = []


class KBStatusResponse(BaseModel):
    uuid: str
    status: str
    total_sources: int = 0
    sources_ready: int = 0
    sources_failed: int = 0
    total_chunks: int = 0
    sources: list[dict] = []


# --- Export / Import ---

KB_EXPORT_FORMAT_VERSION = 1


class KBExportSource(BaseModel):
    source_type: str  # "document" | "url"
    document_uuid: Optional[str] = None
    document_title: Optional[str] = None  # snapshot of SmartDocument.title at export time
    url: Optional[str] = None
    url_title: Optional[str] = None
    custom_name: Optional[str] = None  # user's chosen label, carried across export/import
    content: Optional[str] = None  # cached raw text (for URLs) or document raw_text (for docs)
    crawl_enabled: bool = False
    max_crawl_pages: int = 5
    parent_source_uuid: Optional[str] = None
    crawled_urls: Optional[list[str]] = None


class KBExportPayload(BaseModel):
    format_version: int = KB_EXPORT_FORMAT_VERSION
    exported_at: Optional[str] = None
    title: str
    description: Optional[str] = None
    tags: list[str] = []
    sources: list[KBExportSource] = []


class ImportKBRequest(BaseModel):
    payload: KBExportPayload
    title: Optional[str] = None  # override title on import


class ImportKBResponse(BaseModel):
    uuid: str
    title: str
    imported_sources: int
